resolwe/test_helpers/test_runner.py

```diff
@@ -5,38 +5,298 @@
+    """
+
+    def __init__(self, command, *args, **kwargs):
+        """Initialize instance variables.
+
+        :param command: The django command to be run.
+        :param args: Positional arguments passed to the command.
+        :param kwargs: Keyword arguments passed to the command.
+        """
+        self._django_command = command
+        self._args = args
+        self._kwargs = kwargs
+        self._pid = None
+        super().__init__()
+
+    def __enter__(self):
+        """Set up a context manager for the specified django command.
+
+        The command is executed in a separate process which inherits the
+        python context from the current one.
+        """
+        # In case of parallel execution, this code will be run from a multiprocessing.Pool worker.
+        # Because those are all daemonic, we can't use the normal multiprocessing.Process chrome;
+        # the remaining alternative is plain posix forking, since spawning a process by image name
+        # would be even more messy due to the difficulty in divining the way this process was started.
+        pid = os.fork()
+        if pid == 0:
+            try:
+                self.run()
+            finally:
+                # If sys.exit(0) is used here, django's testing framework will stupidly catch it
+                # as if it was an error, so just commit suicide as quickly as possible. Returning
+                # from the function would lead to chaos and madness because django isn't aware that
+                # we've forked; cleanup handlers would get called twice (from here and from the parent).
+                os.kill(os.getpid(), SIGKILL)
+        else:
+            self._pid = pid
+            return self
+
+    def __exit__(self, *args, **kwargs):
+        """On exiting a context, kill the command and wait for it."""
+        os.kill(self._pid, SIGINT)
+        os.waitpid(self._pid, 0)
+        return False  # propagate exceptions
+
+    def run(self):
+        """Run the django command specified in the constructor."""
+        call_command(self._django_command, *self._args, **self._kwargs)
+
+
+class AtScopeExit(object):
+    """Utility class for calling a function once a context exits."""
+
+    def __init__(self, call, *args, **kwargs):
+        """Construct a context manager and save arguments.
+
+        :param call: The callable to call on exit.
+        :param args: Positional arguments for the callable.
+        :param kwargs: Keyword arguments for the callable.
+        """
+        self.call = call
+        self.args = args
+        self.kwargs = kwargs
+
+    def __enter__(self):
+        """Enter the ``with`` context."""
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        """Exit the context and call the saved callable."""
+        self.call(*self.args, **self.kwargs)
+        return False
+
+
+def _manager_setup():
+    """Execute common setup operations."""
```

Comment:
Can you make the docstring more descriptive? Maybe remove the comment below and include it in the docstring.

Reply:
The comment below turned out to be stale anyway, I neglected to remove it when refactoring the code into a function.